src/coreclr/vm/dllimportcallback.cpp

```diff
@@ -699,8 +699,29 @@ Stub *UMThunkMarshInfo::CompileNExportThunk(LoaderHeap *pLoaderHeap, PInvokeStat
         numRegistersUsed++;
     }
 
+    bool hasReturnBuffer = argit.HasRetBuffArg() || (m_callConv == pmCallConvThiscall && argit.HasValueTypeReturn());
+    bool hasNativeExchangeTypeReturn = false;
+
+    if (hasReturnBuffer)
+    {
+        // If think we have a return buffer, lets make sure that we aren't returning one of the intrinsic native exchange types.
+        TypeHandle returnType = pMetaSig->GetRetTypeHandleThrowing();
+        if (returnType.GetMethodTable()->IsIntrinsicType())
```

Comment:
This looks like its assuming any type under `System.Runtime.InteropServices` and marked as `Intrinsic` is a "native exchange type".

Is that correct? Should we at least have a `Debug.Assert` to check that it is one of the "known" types, just in case some other non-interchange type gets exposed in the future?

Reply:
We do a similar thing in the JIT's support for SIMD types (just check namespace) if I remember correctly. 

And this particular block of code will be deleted when #46238 is merged in, so it's short lived.

Reply:
GetFullyQualifiedNameInfo returns the type name. It is just one extra line to check the type name...

Reply:
I'll add in an assert for the type name to make sure we match the JIT behavior.